go/libkb/secret_store_darwin.go

```diff
@@ -7,138 +7,226 @@ package libkb
+		switch err {
+		case nil:
+			previousSecret = secret
+		case keychain.ErrorItemNotFound, NewErrSecretForUserNotFound(accountName):
+			// We've reached the end of the keychain entries so let's return
+			// the previous secret we found.
+			secret = previousSecret
+			k.updateAccessibility(mctx, account)
+			err = nil
+			mctx.Debug("successfully retrieved secret on attempt: %d", i-1)
+			break Loop
+		default:
+			mctx.Debug("unable to retrieve secret: %v, attempt: %d", err, i)
+		}
+	}
 	if err != nil {
-		m.Debug("KeychainSecretStore.RetrieveSecret(%s) error: %s", accountName, err)
-		k.mobileKeychainPermissionDeniedCheck(m, err)
 		return LKSecFullSecret{}, err
-	}
-	if encodedSecret == nil {
-		m.Debug("KeychainSecretStore.RetrieveSecret(%s) nil encodedSecret", accountName)
+	} else if secret.IsNil() {
 		return LKSecFullSecret{}, NewErrSecretForUserNotFound(accountName)
 	}
+	return secret, nil
+}
 
-	secret, err := base64.StdEncoding.DecodeString(string(encodedSecret))
+func (k KeychainSecretStore) retrieveSecret(mctx MetaContext, account keychainSlottedAccount) (lk LKSecFullSecret, err error) {
+	encodedSecret, err := keychain.GetGenericPassword(k.serviceName(mctx), account.String(),
+		"", k.accessGroup(mctx))
 	if err != nil {
-		m.Debug("KeychainSecretStore.RetrieveSecret(%s) base64.Decode error: %s", accountName, err)
+		k.mobileKeychainPermissionDeniedCheck(mctx, err)
 		return LKSecFullSecret{}, err
+	} else if encodedSecret == nil {
+		return LKSecFullSecret{}, NewErrSecretForUserNotFound(account.name)
 	}
 
-	m.Debug("KeychainSecretStore.RetrieveSecret(%s) got secret, creating lksec", accountName)
-
-	lk, err := newLKSecFullSecretFromBytes(secret)
+	secret, err := base64.StdEncoding.DecodeString(string(encodedSecret))
 	if err != nil {
-		m.Debug("KeychainSecretStore.RetrieveSecret(%s) error creating lksec: %s", accountName, err)
 		return LKSecFullSecret{}, err
 	}
 
-	// Update accessibility
-	k.updateAccessibility(m, accountName.String())
-
-	m.Debug("KeychainSecretStore.RetrieveSecret(%s) success", accountName)
-
-	return lk, nil
+	return newLKSecFullSecretFromBytes(secret)
 }
 
-func (k KeychainSecretStore) ClearSecret(m MetaContext, accountName NormalizedUsername) error {
-	m.Debug("KeychainSecretStore#ClearSecret(%s)", accountName)
+func (k KeychainSecretStore) ClearSecret(mctx MetaContext, accountName NormalizedUsername) (err error) {
+	defer mctx.TraceTimed(fmt.Sprintf("KeychainSecretStore#ClearSecret: accountName: %s", accountName),
+		func() error { return err })()
+
 	if accountName.IsNil() {
-		m.Debug("NOOPing KeychainSecretStore#ClearSecret for empty username")
+		mctx.Debug("NOOPing KeychainSecretStore#ClearSecret for empty username")
 		return nil
 	}
-	query := keychain.NewGenericPassword(k.serviceName(m), string(accountName), "", nil, k.accessGroup(m))
+
+	// Try all slots to fully clear any secrets for this user
+	epick := FirstErrorPicker{}
+	for i := 0; i < maxKeychainItemSlots; i++ {
```

Comment:
Is this slow? Iterating over 50 slots for every user? Even the 99.99% who don't have this issue?

Reply:
Another possibility is just to stop iterating once we hit a slot that was never set?

Reply:
yeah i was worried about it being an issue but in tests it runs very quickly: `KeychainSecretStore#ClearSecret: accountName: username -> ok [time=12.313227ms]` so i thought it was better to be safe and attempt the delete for any possible values since actual `DeleteItem` get false positives of `keychain.ErrorItemNotFound` (we could try to read ahead like `StoreSecret` does but thought it wasn't worth the added complexity)

Reply:
ah ok.  that's fast enough.  Is that on mobile or desktop?

Reply:
desktop, i can double check on mobile but i think it was similar during my manual testing

Reply:
mobile: `KeychainSecretStore#ClearSecret: accountName: test5 -> ok [time=71.138ms] `